File: src/main/java/com/antlr/plugin/psi/AtAction.java
package com.antlr.plugin.psi;

import com.antlr.plugin.parser.PsiElementFactory;
import com.intellij.extapi.psi.ASTWrapperPsiElement;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.antlr.plugin.ANTLRv4TokenTypes;
import com.antlr.plugin.parser.ANTLRv4Parser;
import org.jetbrains.annotations.NotNull;

public class AtAction extends ASTWrapperPsiElement {
    public AtAction(@NotNull ASTNode node) {
        super(node);
    }

    @NotNull
    public String getIdText() {
        PsiElement id = findChildByType(ANTLRv4TokenTypes.getRuleElementType(ANTLRv4Parser.RULE_identifier));

        return id == null ? "<n/a>" : id.getText();
    }

    @NotNull
    public String getActionBlockText() {
        PsiElement actionBlock = findChildByType(ANTLRv4TokenTypes.getRuleElementType(ANTLRv4Parser.RULE_actionBlock));

        if (actionBlock != null) {
            PsiElement openingBrace = actionBlock.getFirstChild();
            PsiElement closingBrace = actionBlock.getLastChild();

            return actionBlock.getText().substring(openingBrace.getStartOffsetInParent() + 1, closingBrace.getStartOffsetInParent());
        }

        return "";
    }

    public static class Factory implements PsiElementFactory {
        public static Factory INSTANCE = new Factory();

        @Override
        public PsiElement createElement(ASTNode node) {
            return new AtAction(node);
        }
    }
}

File: src/main/java/com/antlr/language/psi/AtAction.kt
package com.antlr.language.psi

import com.antlr.language.ANTLRv4Parser
import com.antlr.language.AntlrTokenTypes.getRuleElementType
import com.intellij.extapi.psi.ASTWrapperPsiElement
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement

class AtAction(node: ASTNode) : ASTWrapperPsiElement(node) {
    val idText: String
        get() {
            val id =
                findChildByType<PsiElement?>(getRuleElementType(ANTLRv4Parser.RULE_identifier))

            return if (id == null) "<n/a>" else id.text
        }

    val actionBlockText: String
        get() {
            val actionBlock =
                findChildByType<PsiElement?>(getRuleElementType(ANTLRv4Parser.RULE_actionBlock))

            if (actionBlock != null) {
                val openingBrace = actionBlock.firstChild
                val closingBrace = actionBlock.lastChild

                return actionBlock.text
                    .substring(openingBrace.startOffsetInParent + 1, closingBrace.startOffsetInParent)
            }

            return ""
        }

    class Factory : PsiElementFactory {
        override fun createElement(node: ASTNode): PsiElement {
            return AtAction(node)
        }

        companion object {
            @JvmField
            var INSTANCE: Factory = Factory()
        }
    }
}
